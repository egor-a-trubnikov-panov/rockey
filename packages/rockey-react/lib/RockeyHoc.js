import React from 'react';
import classnames from 'classnames';

import isString from 'lodash/isString';
import isObject from 'lodash/isObject';
import isArray from 'lodash/isArray';

import rule from 'rockey/rule';
import look from './look';

import RockeyHocWithHandlers from './utils/RockeyHocWithHandlers';
import { ROCKEY_MIXIN_HANDLER_KEY } from './handler';

const COMPONENT_EXTENDS = 'COMPONENT_EXTENDS';
const DEFINE_COMPONENT_NAME = 'DEFINE_COMPONENT_NAME';
const WAS_CALLED_AS_REACT_COMPONENT = 'WAS_CALLED_AS_REACT_COMPONENT';

const getCallType = (...args) => {
  if (args.length === 1 && isString(args[0])) {
    return DEFINE_COMPONENT_NAME;
  } else if (isArray(args[0]) || (isObject(args[0]) && args.length === 1)) {
    return COMPONENT_EXTENDS;
  } else {
    return WAS_CALLED_AS_REACT_COMPONENT;
  }
};

let anonymysRockeyCounter = 0;
const childCounter = {};

const createEmtpyCss = name => {
  const css = rule``;
  css.wrapWith(name);
  return css;
};

export const getRockeyHoc = () => {
  const RockeyHoc = (
    BaseComponent,
    {
      displayName,
      parentName,
      css,
      at,
    } = {}
  ) => {
    const name = displayName || `AnonymysRockey${++anonymysRockeyCounter}`;
    let queuedMixins = null;

    const FlexibleRockeyHoc = (...args) => {
      const CALL_TYPE = getCallType(...args);

      switch (CALL_TYPE) {
        case DEFINE_COMPONENT_NAME:
          const childComponentName = args[0];

          return RockeyHoc(BaseComponent, {
            displayName: childComponentName,
            parentName: name,
            css,
            at: DEFINE_COMPONENT_NAME,
          });

        case COMPONENT_EXTENDS:
          const componentCss = rule(...args);

          if (queuedMixins) {
            componentCss.addMixins(queuedMixins);
          }

          if (css) {
            componentCss.addParent(css);
          }

          // TODO: ?
          let childName = null;

          if (parentName && at !== DEFINE_COMPONENT_NAME) {
            if (!childCounter[parentName]) {
              childCounter[parentName] = 0;
            }

            childName = `Child${parentName}-${++childCounter[parentName]}`;
          } else {
            childName = name;
          }

          componentCss.wrapWith(childName);

          return RockeyHoc(BaseComponent, {
            displayName: childName,
            parentName: name,
            css: componentCss,
            at: COMPONENT_EXTENDS,
          });

        case WAS_CALLED_AS_REACT_COMPONENT:
          const props = args[0];

          if (!css) {
            css = createEmtpyCss(name);
          }

          let selector = name;

          if (at === DEFINE_COMPONENT_NAME) {
            selector = parentName;
          }

          const handlers = [];
          Object.keys(css.mixins).forEach(key => {
            const mixin = css.mixins[key];

            if (mixin[ROCKEY_MIXIN_HANDLER_KEY]) {
              handlers.push(mixin);
            }
          });

          if (handlers.length) {
            return (
              <RockeyHocWithHandlers
                css={css}
                selector={selector}
                handlers={handlers}
                BaseComponent={BaseComponent}
                proxy={props}
              />
            );
          } else {
            const classList = css.getClassList(props);
            const className = classnames(classList[selector], props.className);

            return React.createElement(BaseComponent, {
              ...props,
              className,
            });
          }

        default:
          throw new Error('Wrong component call');
      }
    };

    FlexibleRockeyHoc.displayName = `Rockey(${name})`;

    FlexibleRockeyHoc.extends = (displayName, childCss) => {
      if (!css) {
        css = createEmtpyCss(name);
      }

      childCss.addParent(css);

      return RockeyHoc(BaseComponent, {
        displayName: displayName,
        parentName: name,
        css: childCss,
      });
    };

    FlexibleRockeyHoc.look = (...args) => {
      const component = RockeyHoc(BaseComponent, {
        displayName,
        parentName,
        css,
      });

      return look(component, {
        extendBase: false,
      })(...args);
    };

    // TODO: create css object instead of queuedMixns
    FlexibleRockeyHoc.addMixins = mixins => {
      if (css) {
        css.addMixins(mixins);
      } else {
        queuedMixins = mixins;
      }
    };

    return FlexibleRockeyHoc;
  };

  return RockeyHoc;
};

const RockeyHoc = getRockeyHoc();
export default RockeyHoc;
