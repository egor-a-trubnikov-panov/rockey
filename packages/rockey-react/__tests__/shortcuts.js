import isFunction from 'lodash/isFunction';
import RockeyHoc from '../lib';
import htmlTags from '../lib/htmlTags';

describe('shortcuts', () => {
  it('shortcuts', () => {
    htmlTags.forEach(tag => {
      expect(isFunction(RockeyHoc[tag])).toBeTruthy();
    });
  });

  it('set shortcuts is disabled', () => {
    expect(() => {
      RockeyHoc.div = () => {};
    }).toThrow('Override default RockeyHoc shorcuts is disabled');
  });

  it('shortcut callType', () => {
    const Button = RockeyHoc.button`
      color: red;
    `;

    const MyButton = RockeyHoc.button('MyButton')`
      color: red;
    `;

    expect(isFunction(Button)).toBeTruthy();
    expect(Button.name).toEqual('FlexibleRockeyHoc');
    expect(Button.displayName).toEqual('Rockey(ShortcutButton1)');

    expect(isFunction(MyButton)).toBeTruthy();
    expect(MyButton.name).toEqual('FlexibleRockeyHoc');
    expect(MyButton.displayName).toEqual('Rockey(MyButton)');
  });
});
