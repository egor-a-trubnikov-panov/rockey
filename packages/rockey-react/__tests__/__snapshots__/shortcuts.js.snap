// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render shortcuts 1`] = `
<button>
  Button
</button>
`;
